src/test/groovy/graphql/schema/idl/RuntimeWiringTest.groovy

```diff
@@ -248,34 +243,91 @@ class RuntimeWiringTest extends Specification {
+                .strictMode(false)
+                .type(TypeRuntimeWiring.newTypeWiring("Foo").typeResolver(TR1))
+                .type(TypeRuntimeWiring.newTypeWiring("Foo").typeResolver(TR1))
+                .build()
+
+        then:
+        noExceptionThrown()
+        runtimeWiring2.typeResolvers.get("Foo") == TR1
+
+        when:
+        def runtimeWiring3 = RuntimeWiring.newRuntimeWiring()
+                .strictMode(false)
+                .type(TypeRuntimeWiring.newTypeWiring("Foo").enumValues(EVP1))
+                .type(TypeRuntimeWiring.newTypeWiring("Foo").enumValues(EVP1))
+                .build()
+
+        then:
+        noExceptionThrown()
+        runtimeWiring3.getEnumValuesProviders().get("Foo") == EVP1
+
+        when:
+        def runtimeWiring4 = RuntimeWiring.newRuntimeWiring()
+                .strictMode(false)
+                .scalar(Scalars.GraphQLString)
                 .build()
 
         then:
-        runtimeWiring.getDefaultDataFetcherForType("Foo") == DEFAULT_DF
+        noExceptionThrown()
+        runtimeWiring4.scalars.get("String") == Scalars.GraphQLString
 
         when:
-        runtimeWiring = RuntimeWiring.newRuntimeWiring()
+        def typeRuntimeWiring = TypeRuntimeWiring.newTypeWiring("Foo")
+                .strictMode(false)
+                .defaultDataFetcher(DF1)
+                .defaultDataFetcher(DF2)
+                .build()
+
+        then:
+        noExceptionThrown()
+        typeRuntimeWiring.defaultDataFetcher == DF2
+    }
+
+    def "when strict mode on, do not allow default data fetcher redefinition"() {
+        DataFetcher DF1 = env -> "w"
+        DataFetcher DEFAULT_DF = env -> "x"
+        DataFetcher DEFAULT_DF2 = env -> "y"
+
+        // Having a datafetcher and a default for the type is ok
+        when:
+        def runtimeWiring1 = RuntimeWiring.newRuntimeWiring()
                 .type(TypeRuntimeWiring.newTypeWiring("Foo").defaultDataFetcher(DEFAULT_DF))
                 .type(TypeRuntimeWiring.newTypeWiring("Foo").dataFetcher("foo", DF1))
-                .type(TypeRuntimeWiring.newTypeWiring("Foo").dataFetcher("bar", DF2))
-                // we can specifically overwrite it later
-                .type(TypeRuntimeWiring.newTypeWiring("Foo").defaultDataFetcher(DEFAULT_DF2))
                 .build()
 
         then:
-        runtimeWiring.getDefaultDataFetcherForType("Foo") == DEFAULT_DF2
+        runtimeWiring1.getDefaultDataFetcherForType("Foo") == DEFAULT_DF
+
+        // Do not permit redefinition of the default datafetcher
+        when:
+        RuntimeWiring.newRuntimeWiring()
+                .type(TypeRuntimeWiring.newTypeWiring("Foo").defaultDataFetcher(DEFAULT_DF))
+                .type(TypeRuntimeWiring.newTypeWiring("Foo").defaultDataFetcher(DEFAULT_DF2))
+                .build()
 
+        then:
+        def error = thrown(StrictModeWiringException)
+        error.message == "The type Foo already has a default data fetcher defined"
```

Comment:
Covering a new edge case, where the default datafetcher is redefined at the runtime wiring level

We already had a test for when it is redefined inside the type wiring